### ISSUE TITLE
Function addQueryArgs in @wordpress/url should add query args before URL fragment

### ISSUE THREAD

**Issue by davilera** [WordPress/gutenberg#16655]
Function `addQueryArgs` adds query args in the incorrect place when the URL contains a URL fragment. According to [RFC 3986](https://tools.ietf.org/html/rfc3986#section-4.2), query args come before the URL fragment:

`relative-ref  = relative-part [ "?" query ] [ "#" fragment ]`

Therefore, when invoking `addQueryArgs` as follows:

`wp.url.addQueryArgs( 'https://example.com/page/#fragment', { param: 'value' } )`

the **expected result** is:

`https://example.com/page/?param=value#fragment`

but the **invalid result** we get is:

`https://example.com/page/#fragment?param=value`

**Additional context**
- Tested with Gutenberg 6.1.1.

**Comment by paaljoachim** [comment]
Hey @davilera David

I see one merged PR and another that stopped up.
Are you around and do you happen to have the time to finish the PR?

Thanks.

@WordPress/gutenberg-core

**Comment by davilera** [comment]
Hey @paaljoachim. Sure, I'll be happy to help.

[As discussed](https://github.com/WordPress/gutenberg/pull/16656#issuecomment-516420111) with @aduth, we probably need to add a [url-polyfill](https://www.npmjs.com/package/url-polyfill) first and then refactor the code to use `window.URL`. The problem is, I don't know how to properly add this polyfill...

**Comment by gziolo** [comment]
See URL usage in Gutenberg: 

https://github.com/WordPress/gutenberg/blob/d914ff9cdf7f1fdd6fc5362b3390537a755eca0e/packages/url/src/get-query-string.js#L16

See URL polyfill in WordPress Core:

https://github.com/WordPress/wordpress-develop/blob/265180967328e94301fdef82c2fe8c4c30b7b9ca/src/wp-includes/script-loader.php#L131

Can you double check if it is still blocked?

**Comment by davilera** [comment]
That's really nice, @gziolo! I guess somebody already added the polyfill; I should have checked this before.

Okay, then. I'll revisit my PR and fix it.

**Comment by davilera** [comment]
Hi @gziolo, @paaljoachim. This issue has been stalled for quite some time, and I had forgotten about it. After revisiting the `@wordpress/url` package, I noticed the problems still persist. To address the most pressing issues without a full refactor to use the polyfill, I created a [new PR](https://github.com/WordPress/gutenberg/pull/69313). I’d appreciate it if someone could review it and share their thoughts.

**Comment by swissspidy** [comment]
FWIW, the year is 2025, [`URL`](https://developer.mozilla.org/en-US/docs/Web/API/URL) has been around for 10 years, I don't see why this would need any polyfill

**Comment by davilera** [comment]
You're totally right. I meant a full refactor to use `URL` everywhere – at the moment, it's used scarcely.
